lib/tapioca/runtime/reflection.rb

```diff
@@ -183,6 +183,61 @@ def attached_class_of(singleton_class)
 
         T.cast(result, Module)
       end
+
+      sig { params(constant: Module).returns(T::Array[String]) }
+      def file_candidates_for(constant)
+        relevant_methods_for(constant).map do |method|
+          method.source_location&.first
+        end.compact.uniq
+      end
+
+      private
+
+      sig { params(constant: Module).returns(T::Array[UnboundMethod]) }
+      def relevant_methods_for(constant)
+        methods = methods_for(constant).select(&:source_location)
+          .reject { |x| method_defined_by_forwardable_module?(x) }
+
+        return methods unless methods.empty?
+
+        constants_of(constant).flat_map do |const_name|
+          if (const = nested_module?(constant, const_name.to_s))
+            relevant_methods_for(const)
+          else
+            []
+          end
+        end
+      end
+
+      sig { params(constant: Module).returns(T::Array[UnboundMethod]) }
+      def methods_for(constant)
+        modules = [constant, singleton_class_of(constant)]
+        method_list_methods = [
+          PUBLIC_INSTANCE_METHODS_METHOD,
+          PROTECTED_INSTANCE_METHODS_METHOD,
+          PRIVATE_INSTANCE_METHODS_METHOD,
+        ]
+
+        modules.product(method_list_methods).flat_map do |mod, method_list_method|
+          method_list_method.bind_call(mod, false).map { |name| mod.instance_method(name) }
+        end
+      end
+
+      sig { params(parent: Module, name: String).returns(T.nilable(Module)) }
+      def nested_module?(parent, name)
```

Comment:
nitpick: having `?` postfix but not returning a boolean feels kinda surprising?

Reply:
Done!